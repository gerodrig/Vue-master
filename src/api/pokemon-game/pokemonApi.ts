import axios from 'axios';

const pokemonApiGraphql = axios.create({
    baseURL: 'https://beta.pokeapi.co/graphql/v1beta',
    headers: {
        'Content-Type': 'application/json',
    },
});

export default pokemonApiGraphql;