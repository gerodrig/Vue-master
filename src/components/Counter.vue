<template>
    <div class="mx-auto text-center debug">
        <h1 class="text-4xl">Counter</h1>
        <p data-testid="counter" class="my-5 text-2xl"> {{ count }} <sup>2</sup> = {{ squared }}</p>
        <div class="buttons">
        <button data-testid="decrement" class="px-4 py-2 font-bold text-white bg-yellow-500 rounded hover:bg-yellow-700" @click="decrement">-1</button>
        <button data-testid="increment" class="px-4 py-2 ml-5 font-bold text-white bg-green-500 rounded hover:bg-green-700" @click="increment">+1</button>
        </div>
    </div>
</template>


<script lang="ts">
    export default {
        name: 'Counter',
        data() {
            return {
                count: 0
            }
        },
        methods: {
            increment() {
                this.count++
            },
            decrement() {
                this.count--
            }
        },
        computed: {
            squared() {
                return this.count * this.count
            }
        },
    }
</script>