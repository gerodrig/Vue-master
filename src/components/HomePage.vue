<template>
  <div class="flex flex-col justify-between min-h-screen bg-gray-100">
    <Header />
    <main class="container p-4 mx-auto">
      <Logo />
      <div class="grid grid-cols-1 gap-6 md:grid-cols-2 lg:h-96 lg:grid-cols-3">

      <router-link v-for="project in data" :key="project.id" :to="project.path">
        <Card :data="project" />
      </router-link>

      </div>
    </main>
    <Footer title="Gerardo Rodriguez" />
  </div>

<!-- //example to keep alive -->
  <!-- <router-view v-slot="{ Component }">
   <keep-alive>
     <component :is="Component" />
    </keep-alive>
  </router-view> -->
</template>

<script lang="ts">
import { Footer, Header, Logo, Card } from '@components/shared';

interface Project {
  id: number;
  name: string;
  path: string;
  description?: string;
}

export default {
  name: 'HomePage',
  components: { Footer, Header, Logo, Card },
  props: {
    msg: String,
    data: {
      type: Array as () => Project[],
      default: () => [],
      validator: (value: Project[]) => {
        return value.every((item: Project) => {
          return (
            item.hasOwnProperty('id') &&
            item.hasOwnProperty('name') &&
            item.hasOwnProperty('path')
          );
        });
      },
    },
  },
};
</script>
