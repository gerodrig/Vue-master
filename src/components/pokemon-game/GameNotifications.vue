<template>
    <div class="absolute top-36 right-12 md:right-36 md:top-1/3 xl:right-24 xl:top-1/3 2xl:right-1/4 2xl:w-96 2xl:h-96">
        <div v-if="showAnswer" class="flex flex-col items-center justify-center w-full h-full max-w-md px-6 py-4 mx-auto my-4 font-semibold text-center text-gray-800 bg-white border-2 border-blue-500 rounded-lg shadow-md ">
                        <h1 v-if="won" class="text-4xl font-bold text-center text-yellow-500">{{message}}</h1>
                        <h1 v-else="won" class="text-3xl font-bold text-center text-red-500">{{message}}</h1>

                        <!-- //button with a pokemon style -->
                        <button class="px-4 py-2 mt-10 font-bold text-white bg-red-500 rounded hover:bg-red-700" @click="$emit('resetGame')">Restart Game</button>
                    </div> 
    </div>
                
</template>

<script lang='ts'>
import { defineComponent, toRef } from 'vue';
export default defineComponent({
  name: 'GameNotifications',
  props: {
    message: {
      type: String,
      required: true,
    },
    won: {
      type: Boolean,
      required: true,
    },
    showAnswer: {
      type: Boolean,
      required: true,
    },
  },
  setup(props) {
    const showAnswer = toRef(props, 'showAnswer');
    const won = toRef(props, 'won');
    const message = toRef(props, 'message');
    return {
      showAnswer,
        won,
        message,
    };
  },
});
</script>