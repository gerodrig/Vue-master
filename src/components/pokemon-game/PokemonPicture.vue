
<template>
  <div :class="className">
    <img
      v-show="id != 0"
      :src="imgSrc"
      alt="`Pokemon ?`"
      :class="{ 'hidden-pokemon': !showPokemon, fadeIn: showPokemon }"
    />
    <!-- <button @click="toggleReveal">Toggle Hide</button> -->
  </div>
</template>

<script lang="ts">
import { defineComponent, computed, toRef } from 'vue';
export default defineComponent({
  name: 'PokemonPicture',
  props: {
    pokemonId: {
      type: Number || undefined,
      required: true,
    },
    showPokemon: {
      type: Boolean,
      default: false,
    },
    defaultPath: {
      type: String,
      default:
        'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/',
    },
    class: {
      type: String,
      default: 'container flex flex-col justify-center my-10 h-52 debug',
    },
  },
  setup(props) {

    const showPokemon = toRef(props, 'showPokemon');
    const id = toRef(props, 'pokemonId');
    const className = toRef(props, 'class');

    const imgSrc = computed(() => {
      const validId =
        props.pokemonId >= 1 && props.pokemonId <= 649 ? props.pokemonId : 0;
      if (validId === 0) {
        return 'https://upload.wikimedia.org/wikipedia/commons/3/3b/MissingNo.svg';
      } else {
        return `${props.defaultPath}${validId}.svg`;
      }
    });

    

    return {
        id,
      showPokemon,
      imgSrc,
      className,
    };
  },
});
</script>

<style scoped>
.fadeIn {
  animation: fadeIn 0.6s;
}

.hidden-pokemon {
  filter: brightness(0);
}

@keyframes fadeIn {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}

/* :style="{objectFit: 'cover', maxWidth: '100%', width: '100%'}" */

img {
object-fit: cover;
max-width: 100%;
width: 100%;
  height: 400px;
  right: 32%;
  user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-user-drag: none;
  -webkit-user-select: none;
}

@media screen and (max-width: 768px) {
  img {
    height: 200px;
  }
    
}
</style>
