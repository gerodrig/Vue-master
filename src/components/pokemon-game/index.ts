export { default as PokemonPicture } from './PokemonPicture.vue';
export { default as PokemonOptions } from './PokemonOptions.vue';
export { default as GameNotifications } from './GameNotifications.vue';