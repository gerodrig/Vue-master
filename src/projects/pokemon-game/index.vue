<template>

  <div
    class="relative w-full h-[80vh] md:h-screen bg-cover bg-[url(@/assets/pokemon-game/background.png)] transform -translate-y-36"
  >
    <h1 v-if="!options">Loading please wait...</h1>

    <PokemonPicture
      v-else
      class="absolute left-28 top-64 md:left-56 md:top-56 xl:left-1/4 xl:top-1/3"
      :pokemon-id="answer.id"
      :show-pokemon="showPokemon"
    />

    <GameNotifications
      :message="message"
      :won="won"
      :show-answer="showAnswer"
      @reset-game="resetGame"
    />

  </div>
  <div class="-mt-36">
    <PokemonOptions
      :options="options"
      :disabled="showAnswer"
      @chosen-pokemon="checkAnswer"
    />
  </div>
</template>

<script lang="ts">
import { defineComponent } from "vue";
import {
  PokemonPicture,
  PokemonOptions,
  GameNotifications,
} from "@components/pokemon-game";
import usePokemon from "@/composables/pokemon-game/usePokemon";

export default defineComponent({
  components: {
    PokemonPicture,
    PokemonOptions,
    GameNotifications,
  },
  name: "Pokemon",
  setup() {

    const {
      options,
      answer,
      showPokemon,
      showAnswer,
      message,
      won,
      checkAnswer,
      resetGame,
    } = usePokemon();

    return {
      options,
      answer,
      showPokemon,
      showAnswer,
      message,
      won,
      checkAnswer,
      resetGame,
    };
  },
});
</script>
