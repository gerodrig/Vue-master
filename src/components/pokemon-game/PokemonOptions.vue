<template>
    <div class="bg-black opacity-80">
        <div class="flex items-center mx-auto justify-center max-w-[800px]">
            <ul class="flex-wrap justify-center py-6 text-blue-700 list-none md:flex">
            <li
                v-for="{ name, id } in options"
                @click="$emit('chosenPokemon', id)"
                :class="{ 'opacity-50': disabled }"
                :style="{ 'pointer-events': disabled ? 'none' : 'auto' }"
                :aria-disabled="disabled"
                :key="id">{{
                name.toLocaleUpperCase()
                }}</li>
            </ul>
        </div>
    </div>
</template>

<script lang="ts">
import { defineComponent, toRef } from 'vue';
import { Pokemon } from '@interfaces/pokemon-game/index';

export default defineComponent({
  name: 'PokemonOptions',
  props: {
    options: {
      type: Array as () => Pokemon[],
      required: true,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
  },
  setup(props) {
    
    const options = toRef(props, 'options');
    const disabled = toRef(props, 'disabled');

    //upercase first letter
    options.value.forEach((option) => {
      option.name = option.name[0].toUpperCase() + option.name.slice(1);
    });
    return {
      options,
      disabled,
    };
  },
});
</script>

<style scoped>
li {
  @apply bg-white rounded-md border border-gray-300 cursor-pointer mb-3 w-96 hover:bg-gray-100 text-center py-2 md:mr-3;
}
</style>
