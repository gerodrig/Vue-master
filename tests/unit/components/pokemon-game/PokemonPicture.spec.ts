import { describe, expect, it } from 'vitest';
import { shallowMount } from '@vue/test-utils';
import {PokemonPicture} from '../../../../src/components/pokemon-game/';


describe('PokemonPicture Component', () => {

  it('should match the snapshot', () => {
    
    const wrapper = shallowMount(PokemonPicture, {
        props: {
            pokemonId: 1,
            showPokemon: false,
        },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should show the hidden image and pokemon number 100', () => {

    const wrapper = shallowMount(PokemonPicture, {
        props: {
            pokemonId: 100,
            showPokemon: false,
        },
    });

    const img = wrapper.find('img');

    expect(img.classes('hidden-pokemon')).toBe(true);
    expect(img.attributes('src')).toBe('https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/100.svg');
  });

  it('should show missigno pokemon if value passed is 0',() => {

    const wrapper = shallowMount(PokemonPicture, {
        props: {
            pokemonId: 0,
            showPokemon: false,
        },
    });

    const img = wrapper.find('img');

    expect(img.classes('hidden-pokemon')).toBe(true);
    expect(img.attributes('src')).toBe('https://upload.wikimedia.org/wikipedia/commons/3/3b/MissingNo.svg');
  });

  it('should show the pokemon image if showPokemon is true', () => {
    
        const wrapper = shallowMount(PokemonPicture, {
            props: {
                pokemonId: 100,
                showPokemon: true,
            },
        });
    
        const img = wrapper.find('img');
    
        expect(img.exists()).toBe(true);
        expect(img.classes('fadeIn')).toBe(true);
        expect(img.classes('hidden-pokemon')).toBe(false);
        expect(img.attributes('src')).toBe('https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/100.svg');
  });
});