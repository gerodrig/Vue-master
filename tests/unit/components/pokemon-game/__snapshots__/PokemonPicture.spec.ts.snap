// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PokemonPicture Component > should match the snapshot 1`] = `
"<div data-v-de2a09f6=\\"\\" class=\\"container flex flex-col justify-center my-10 h-52 debug\\"><img data-v-de2a09f6=\\"\\" src=\\"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg\\" alt=\\"\`Pokemon ?\`\\" class=\\"hidden-pokemon\\"><!-- <button @click=\\"toggleReveal\\">Toggle Hide</button> -->
</div>"
`;
